dbms/src/Interpreters/Aggregator.cpp

```diff
@@ -2572,6 +2139,170 @@ void Aggregator::setCancellationHook(CancellationHook cancellation_hook)
+    , aggregator(aggregator_)
+    , data(data_)
+    , final(final_)
+    , concurrency(concurrency_)
+{
+    assert(concurrency > 0);
+    if (!data.empty())
+    {
+        is_two_level = data[0]->isTwoLevel();
+        if (is_two_level)
+        {
+            for (size_t i = 0; i < concurrency; ++i)
+                two_level_parallel_merge_data.push_back(std::make_unique<BlocksList>());
+        }
+        else
+        {
+            // for single level, concurrency must be 1.
+            RUNTIME_CHECK(concurrency == 1);
+        }
+
+        /// At least we need one arena in first data item per concurrency
+        if (concurrency > data[0]->aggregates_pools.size())
+        {
+            Arenas & first_pool = data[0]->aggregates_pools;
+            for (size_t j = first_pool.size(); j < concurrency; ++j)
+                first_pool.emplace_back(std::make_shared<Arena>());
+        }
+    }
+}
+
+Block MergingBuckets::getHeader() const
+{
+    return aggregator.getHeader(final);
+}
+
+Block MergingBuckets::getData(size_t concurrency_index)
+{
+    assert(concurrency_index < concurrency);
+
+    if (unlikely(data.empty()))
+        return {};
+
+    FAIL_POINT_TRIGGER_EXCEPTION(FailPoints::random_aggregate_merge_failpoint);
+
+    return is_two_level ? getDataForTwoLevel(concurrency_index) : getDataForSingleLevel();
+}
+
+Block MergingBuckets::getDataForSingleLevel()
+{
+    assert(!data.empty());
+
+    AggregatedDataVariantsPtr & first = data[0];
+
+    if (current_bucket_num == -1)
+    {
+        ++current_bucket_num;
+
+        if (first->type == AggregatedDataVariants::Type::without_key)
+        {
+            aggregator.mergeWithoutKeyDataImpl(data);
+            single_level_blocks = aggregator.prepareBlocksAndFillWithoutKey(
+                *first,
+                final);
+            return popBlocksListFront(single_level_blocks);
+        }
+    }
+
+    Block out_block = popBlocksListFront(single_level_blocks);
+    if (likely(out_block))
+    {
+        return out_block;
+    }
+
+    if (current_bucket_num > 0)
+        return {};
+
+    if (first->type == AggregatedDataVariants::Type::without_key)
+        return {};
+
+    ++current_bucket_num;
+
+#define M(NAME)                                                                 \
+    case AggregationMethodType(NAME):                                           \
+    {                                                                           \
+        aggregator.mergeSingleLevelDataImpl<AggregationMethodName(NAME)>(data); \
+        break;                                                                  \
+    }
+    switch (first->type)
+    {
+        APPLY_FOR_VARIANTS_SINGLE_LEVEL(M)
+    default:
+        throw Exception("Unknown aggregated data variant.", ErrorCodes::UNKNOWN_AGGREGATED_DATA_VARIANT);
+    }
+#undef M
+    single_level_blocks = aggregator.prepareBlocksAndFillSingleLevel(*first, final);
```

Comment:
Why not move the merge code to L2201, together with the `without_key` case?

Reply:
yes, you are right, updated.